File: MATLAB/Method/Statistical_methods/Time_domain/PGA/PGA.m
function [Pulse_shape_discrimination_factor] = PGA(Pulse_signal, t)

%--------------------------------------------------------------------------
% Pulse Gradient Analysis (PGA) method for pulse shape discrimination.
%
% Discriminates particle types by analyzing the decay rate of pulse signals
% immediately after their peak. Computes the gradient between the peak and
% a fixed time delay to capture differences in decay characteristics.
%
% Input:
%   Pulse_signal: Matrix of pulse signals to analyze [n_signals × signal_length]
%   t: Time delay after peak for gradient calculation (default: 20 samples)
%
% Output:
%   Pulse_shape_discrimination_factor: Signal gradients after peak
%                                    [n_signals × 1]
%
% Algorithm:
%   1. Find peak amplitude and position for each signal
%   2. Sample amplitude at fixed delay t after peak
%   3. Calculate gradient: g = (A(t_peak + t) - A_peak) / t
%   where A is amplitude, t_peak is peak position
%
% Note:
%   - Gradient calculation requires t samples after peak
%   - Shorter t focuses on initial decay
%   - Longer t captures more of decay curve
%   - Error if peak + t exceeds signal length
%
% Reference:
% D'Mellow, Bob, et al. "Digital discrimination of neutrons and γ-rays in
% liquid scintillators using pulse gradient analysis." Nuclear Instruments
% and Methods in Physics Research Section A: Accelerators, Spectrometers,
% Detectors and Associated Equipment 578.1 (2007): 191-197.
%--------------------------------------------------------------------------

% Set default time delay
if nargin < 2
    t = 20;  % Samples after peak
end

% Initialize output array
Num_signal = size(Pulse_signal, 1);
Pulse_shape_discrimination_factor = zeros(Num_signal, 1);

% Process signals in parallel
parfor i = 1:Num_signal
    % Find peak characteristics
    [Max_value, Max_index] = max(Pulse_signal(i, :));  % [amplitude, position]
    
    % Calculate delayed sample position
    Second_sample_index = Max_index + t;
    
    % Validate delayed position
    if Second_sample_index > size(Pulse_signal, 2)
        error('Delayed sample at peak+%d exceeds signal length for signal %d.', t, i);
    end
    
    % Get delayed amplitude
    Second_sample_value = Pulse_signal(i, Second_sample_index);
    
    % Compute decay gradient
    Pulse_shape_discrimination_factor(i) = (Second_sample_value - Max_value) / t;
end

end

File: Method/Statistical_methods/Time_domain/pga.py
"""
Pulse Gradient Analysis (PGA) method for pulse shape discrimination.

The method computes discrimination factors by calculating the gradient 
between the pulse peak and a fixed time point after the peak, providing
a simple but effective measure of pulse shape differences.

Reference:
- D’Mellow, Bob, M. D. Aspinall, R. O. Mackin, Malcolm J. Joyce, and A. J. Peyton.
  "Digital discrimination of neutrons and γ-rays in liquid scintillators using pulse gradient analysis."
  Nuclear Instruments and Methods in Physics Research Section A: Accelerators, Spectrometers,
  Detectors and Associated Equipment 578, no. 1 (2007): 191-197.
"""

import numpy as np


def get_psd_factor(
        pulse_signal: np.ndarray,
        t: int = 20
) -> np.ndarray:
    """
    Calculate PSD factors using Pulse Gradient Analysis.
    
    Args:
        pulse_signal: Input signals array of shape (N, L) where N is the number
                     of pulses and L is the length of each pulse
        t: Number of samples after the peak to use for gradient calculation
           (default: 20)
    
    Returns:
        numpy.ndarray: Array of PSD factors for each input pulse, computed as
                      (V[peak + t] - V[peak]) / t. NaN values indicate invalid
                      calculations (e.g., peak + t beyond signal length)
    
    Note:
        The gradient is calculated as a simple finite difference over t samples,
        normalized by the time interval to give a rate of change.
    """
    num_signals = pulse_signal.shape[0]
    discrimination_factors = np.full(num_signals, np.nan)

    for i, signal in enumerate(pulse_signal):
        # Find the peak value and its index
        max_index = np.argmax(signal)   # Adjust to MATLAB-style indexing
        max_value = signal[max_index ]  # Adjust the access to match Python's 0-indexing

        # Calculate the index for the sample t samples after the peak
        second_sample_index = max_index + t

        # Check if the second sample index is within bounds
        if second_sample_index < signal.shape[0]:
            second_sample_value = signal[second_sample_index]
            discrimination_factors[i] = (second_sample_value - max_value) / t
        else:
            # If out-of-bounds, assign NaN for this signal
            discrimination_factors[i] = np.nan

    return discrimination_factors

